ENH Introduce `PairwiseDistancesReduction` and `PairwiseDistancesArgKmin` (feature branch)

jjerphan opened the pull request:
Supersedes #21462.
Relates to #22587.

Task list as mentioned in https://github.com/scikit-learn/scikit-learn/pull/21462#issuecomment-993862788:

---
  - [x] Address the last reviews here as mentioned in https://github.com/scikit-learn/scikit-learn/pull/21462#issuecomment-992581822
  - [x] Finalize the sub-PR to use a class method .compute jjerphan/scikit-learn#5 for the small refactoring suggested by @thomasjpfan.
  - [x] Refactor heap routines #21987.
  - [x] https://github.com/scikit-learn/scikit-learn/pull/22134 : high-level PR for the feature branch
  - [x] Create the following branches which target this branch In this order (the proposal above are being merged and reordered here):
    - [x] https://github.com/scikit-learn/scikit-learn/pull/22064
    - [x] https://github.com/scikit-learn/scikit-learn/pull/22065
    - [x] https://github.com/scikit-learn/scikit-learn/pull/22288
 - [x] Merge all of them
 - [x] Merge `pairwise-distances-argkmin` in `main`
 ---

### Results

Using [this script](https://gist.github.com/jjerphan/7db77ec6eb081c4859417eeaca49f769):

On scikit-learn `1.0`:

![speed_up_100000_100000_log](https://user-images.githubusercontent.com/13029839/155144240-f5cd7cfe-ebdf-4d71-be9a-5ce9e4d62403.svg)
![speed_up_1000000_10000_log](https://user-images.githubusercontent.com/13029839/155144242-6c041729-154b-47aa-9069-3a7d26deef5a.svg)


This PR:

![1](https://user-images.githubusercontent.com/13029839/155096008-496d6b2d-d65a-4848-8eb9-b8bc1cd83d50.svg)
![2](https://user-images.githubusercontent.com/13029839/155096010-d143649b-3904-4d80-b3d0-5017724d19ad.svg)

See https://github.com/scikit-learn/scikit-learn/pull/21462 for reference.

jjerphan committed:
MAINT Introduce Pairwise Distances Reductions private submodule  (#22064)

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

jjerphan committed:
MAINT Introduce FastEuclideanPairwiseArgKmin  (#22065)

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

thomasjpfan commented:
I am okay with merging this into `main` and have a follow up PR hook everything up. This was the approach we took with the loss module: https://github.com/scikit-learn/scikit-learn/pull/20567

lorentzenchr commented:
As this code is reviewed, we can merge it anytime.
@jjerphan Is there anything else you would like to have in this branch (where separate PR to main would be more difficult)?

jjerphan committed:
Merge branch 'main' into pairwise-distances-argkmin

jjerphan committed:
fixup! Merge branch 'main' into pairwise-distances-argkmin

Remove duplicated Bunch

thomasjpfan commented:
> I was planning to stick to the plan, doing another sub-PR as initially planned not to have "dead-code" integrated.

No strong opinion either. We can stick with the original plan.

This means we need one more PR into `pairwise-distances-argkmin` to hook everything up?

jjerphan commented:
> No strong opinion either. We can stick with the original plan.

I believe this is the best approach for posterity.

> This means we need one more PR into pairwise-distances-argkmin to hook everything up?

That's it. I'll do it tomorrow.

ogrisel commented:
Also `test_pairwise_distances_argkmin` has a special branch for `"haversine"` but `"haversine"` is not part of the `CDIST_PAIRWISE_DISTANCES_REDUCTION_COMMON_METRICS` list.

Either it should be added to the list or that branch should be removed from the test.

https://app.codecov.io/gh/scikit-learn/scikit-learn/compare/22134/diff#D5L329

jjerphan committed:
Merge branch 'main' into pairwise-distances-argkmin

jjerphan committed:
Merge branch 'main' into pairwise-distances-argkmin

jjerphan committed:
Merge branch 'main' into pairwise-distances-argkmin

jjerphan committed:
MAINT Plug `PairwiseDistancesArgKmin` as a back-end (#22288)

* Forward pairwise_dist_chunk_size in the configuration

* Flip finalized results for PairwiseDistancesArgKmin

The previous would have made the code more complex
by introducing some boilerplate for the interface plugs.

Having it this way actually simplifies the code.

This also removes the haversine branch for
test_pairwise_distances_argkmin

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

* Plug PairwiseDistancesArgKmin as a back-end

* Adapt test accordingly

* Add whats_new entry

* Change input validation order for kneighbors

* Remove duplicated test_neighbors_distance_metric_deprecation

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

* Adapt the documentation

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

* Add mahalanobis case to test fixtures

* Correct whats_new entry

* CLN Remove unneeded private metric attribute

This was needed when 'fast_sqeuclidean' and 'fast_euclidean'
were present to choose the best implementation based on the user
specification.

Those metric have been removed since then, making this attribute
useless.

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

* TST Assert FutureWarning instead of DeprecationWarning in
test_neighbors_metrics

* MAINT Add use_pairwise_dist_activate to scikit-learn config

* TST Add a test for the 'brute' backends' results' consistency

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

* fixup! MAINT Add use_pairwise_dist_activate to scikit-learn config

* fixup! fixup! MAINT Add use_pairwise_dist_activate to scikit-learn config

* TST Filter FutureWarning for WMinkowskiDistance

* MAINT pin numpydoc in arm for now (#22292)

* fixup! TST Filter FutureWarning for WMinkowskiDistance

* Revert keywords arguments removal for the GEMM trick for 'euclidean'

* MAINT pin max numpydoc for now (#22286)

* Add 'haversine' to CDIST_PAIRWISE_DISTANCES_REDUCTION_COMMON_METRICS

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

* fixup! Add 'haversine' to CDIST_PAIRWISE_DISTANCES_REDUCTION_COMMON_METRICS

* Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

* MAINT Document some config parameters for maintenance

Also rename one of them.

Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

* FIX Support and test one of 'sqeuclidean' specification

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

* FIX Various typos fix and correct haversine

'haversine' is not supported by cdist.

* Directly use get_config

* CLN Apply comments from review

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
Co-authored-by: Jérémie du Boisberranger
<jeremiedbb@users.noreply.github.com>

* Motivate swapped returned values

Co-authored-by: Jérémie du Boisberranger
<jeremiedbb@users.noreply.github.com>
Co-authored-by: Thomas J. Fan <thomasjpfan@gmail.com>

* TST Remove mahalanobis from test fixtures

Co-authored-by: Jérémie du Boisberranger <jeremiedbb@users.noreply.github.com>

* MNT Add comment regaduction functions' signatures

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>
Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

* TST Complete test for `pairwise_distance_{argmin,argmin_min}` (#22371)

* DOC Add sub-pull requests to the whats_new entry

jjerphan committed:
Merge branch 'main' into pairwise-distances-argkmin

jjerphan commented:
> Is there anything left or can those 2k loc be merged?

Nothing that I can remember of.

lorentzenchr commented:
> > Is there anything left or can those 2k loc be merged?
> 
> Nothing that I can remember of.

In that case, let's wait one or a few days, if anybody has any further comment. Eventually
1. put it out of draft modus
2. the honor is yours @jjerphan (if you like and are available to push the merge button)

lorentzenchr committed:
Merge branch 'main' into pairwise-distances-argkmin

thomasjpfan commented on sklearn/metrics/pairwise.py:
```diff
@@ -576,12 +578,23 @@ def _euclidean_distances_upcast(X, XX=None, Y=None, YY=None, batch_size=None):
     return distances
 
 
+# start is specified in the signature of `_argmin_min_reduce`
+# and of `_argmin_reduce` but is not used.
+# This is because the higher order `pairwise_distances_chunked`
+# function needs reduction functions that are passed as argument
+# to have a two arguments signature.
```
I prefer to have this comment be in the body of `_argmin_min_reduce` and `_argmin_reduce`? There is a change the comment will get loss if `_argmin_reduce` gets moved around.

lorentzenchr replied:
Fixed with https://github.com/scikit-learn/scikit-learn/pull/22134/commits/ca6f379182e96f024193390b213cc4c90fc56598.

thomasjpfan reviewed (approved):
Minor nit, otherwise LGTM

lorentzenchr committed:
DOC place comment inside functions

lorentzenchr committed:
DOC move up whatsnew entry

lorentzenchr reviewed (approved):
LGTM

GaelVaroquaux commented:
Woot!! This is a big deal!

Congratulations to everyone involved!!!

GaelVaroquaux commented:
Thanks to @jjerphan for the hard work, and to the reviewers!!

jjerphan commented:
_Coming back from holidays_ :mountain:

---

Thanks to everyone involved! This was a big piece which paves the way to smaller PRs, which reviews hopefully would be eased.

> Thanks to @jjerphan for the hard work, and to the reviewers!!

You're welcome.
